### PR TITLE
Move mbedtls crypto modules

### PR DESCRIPTION
**Pull request by ronald-cron-arm**
## Description
Repo split preparation, move Mbed TLS crypto modules to tf-psa-crypto.
Companion mbedtls-framework PR: https://github.com/Mbed-TLS/mbedtls-framework/pull/32
Fix https://github.com/Mbed-TLS/mbedtls/issues/9263

## PR checklist
- [x] **changelog** not required, we will have one change log for the whole repo split work
- [x] **3.6 backport** https://github.com/Mbed-TLS/mbedtls/pull/9383
- [x] **2.28 backport** nothing to backport
- [x] **tests** no additional test required

### CODE REVIEW

File: tests/scripts/check_files.py
Comment (davidhorstmann-arm) [inline tests/scripts/check_files.py:373]: Since we no longer have a dedicated `3rdparty` directory, it'd be nice to have a default of "this cannot have a different license" (i.e. safe by default) in case we add other official drivers in future.

I believe this could be achieved like so:
```suggestion
        r'tf-psa-crypto/drivers/(?=(everest)/.*)',
```

Would also need to change the description to something like:
```python
# Exempt third-party drivers which may be under a different license
```

File: library/.gitignore
Comment (ronald-cron-arm) [reply library/.gitignore:7]: They will be back at some point with one error.c/version_features.c generated in mbedtls and TF-PSA-Crypto thus it seems fine to let then here.

File: tf-psa-crypto/drivers/builtin/CMakeLists.txt
Comment (ronald-cron-arm) [reply tf-psa-crypto/drivers/builtin/CMakeLists.txt:1]: Without this, in an out-of-source-tree build the `src` directory is not created in the build tree and the generation of error.c and version_features.c in drivers/builtin/src fails. Thus that's why I've added it but if you have another better way, I am happy to look at it.

File: tests/scripts/check_files.py
Comment (ronald-cron-arm) [reply tests/scripts/check_files.py:373]: Yes good idea, thanks.

File: tests/scripts/check-generated-files.sh
Comment (ronald-cron-arm) [reply tests/scripts/check-generated-files.sh:136]: Yes you are right. I've fixed that and done a bit more. Not in an additional commit though as I spotted that before you posted your review.

File: tests/scripts/all.sh
Comment (ronald-cron-arm) [reply tests/scripts/all.sh:4632]: This one is a bit subtle. We need to do it like that to match the `library/%` target in the root Makefile. As we plan to abandon the make build system, I haven't tried to improve it.

**Comment by ronald-cron-arm** [comment]
@davidhorstmann-arm thanks for the very quick review. The CI was passing but I was still changing things locally when you pushed your review thus the force-pushed, sorry about that. The diff between https://github.com/Mbed-TLS/mbedtls/pull/9340/commits/499a436c5ab396e45c93c5f6e859f77ef41ff89c and https://github.com/Mbed-TLS/mbedtls/pull/9340/commits/f87e344bdf069feb72ac1b475b1b3243cbdc1570 is not too big though. I hope it will be okay with you. I've added 3 commits to address some of your comments. One of your comment has been addressed as part of the rework I was doing.
I am checking the CI now and will rebase later to resolve the conflicts.

**Comment by ronald-cron-arm** [comment]
The OpenCI has run successfully, rebasing now.

File: tf-psa-crypto/drivers/builtin/CMakeLists.txt
Comment (davidhorstmann-arm) [reply tf-psa-crypto/drivers/builtin/CMakeLists.txt:1]: Hmm. It's not worth holding this PR up over but I might have a look and see if it can be made neater once this is merged.

**Review by davidhorstmann-arm — DISMISSED** [review]
Rebase and changes LGTM, thanks!

**Comment by davidhorstmann-arm** [comment]
I expect this will need a few rebases before it gets merged, though. For example, it'll need rebasing after #9214 is merged.

**Comment by ronald-cron-arm** [comment]
> I expect this will need a few rebases before it gets merged, though. For example, it'll need rebasing after #9214 is merged.

Yes that's likely.

**Comment by ronald-cron-arm** [comment]
@davidhorstmann-arm @tom-daubney-arm  a lot of conflicts but the rebase is still straightforward thus I let it like that for the time being.

**Review by tom-daubney-arm — APPROVED** [review]
LGTM  now, thanks! I will await the rebase.

**Comment by ronald-cron-arm** [comment]
@tom-daubney-arm @davidhorstmann-arm I've just rebased against the head of development. I had to resolve only two conflicts manually:
- commit "Move crypto modules", the file `library/ecp_internal_alt.h` was moved by the commit but deleted by #9271. Resolution: "git rm" of the file.
- commit "Adapt check-generated-files.sh": framework conflict resolved by rebasing on top of `main` and the framework companion PR32 has been updated accordingly.

Let's see what the CI says now.

**Comment by ronald-cron-arm** [comment]
@tom-daubney-arm OpenCI passed and I have then updated the framework submodule following the merge of #32. It is ready for re-review I believe.

**Review by tom-daubney-arm — APPROVED** [review]
Rebase (and fw update) LGTM, thanks!

**Review by minosgalanakis — APPROVED** [review]
LGTM (Icremental review on [post rebase changes](https://github.com/Mbed-TLS/mbedtls/compare/0eb5e9ef07ca954fd1d59941dc890ac0232695bd..080ab4f4ad2d38a1feabf9e23051658fd11a3daf))

**Comment by ronald-cron-arm** [comment]
> LGTM (Icremental review on [post rebase changes](https://github.com/Mbed-TLS/mbedtls/compare/0eb5e9ef07ca954fd1d59941dc890ac0232695bd..080ab4f4ad2d38a1feabf9e23051658fd11a3daf))

Thanks @minosgalanakis

**Comment by ronald-cron-arm** [comment]
@minosgalanakis please have a look to the very partial backport as well, thanks.

**Comment by minosgalanakis** [comment]
> @minosgalanakis please have a look to the very partial backport as well, thanks. #9383

Is that a backport or just a single commit mirroing [1992c9122f](https://github.com/Mbed-TLS/mbedtls/pull/9340/commits/1992c9122fca78dc3c6e282c52d1b1f2019eb9af) ?

**Comment by ronald-cron-arm** [comment]
> > @minosgalanakis please have a look to the very partial backport as well, thanks. #9383
> 
> Is that a backport or just a single commit mirroing [1992c9122f](https://github.com/Mbed-TLS/mbedtls/pull/9340/commits/1992c9122fca78dc3c6e282c52d1b1f2019eb9af) ?

It is the only thing to backport to me. All other changes in #9340 are migration of C modules and their consequences which are not relevant in 3.6.

File: library/Makefile
Comment (gilles-peskine-arm) [inline library/Makefile:402]: On Windows, crypto objects in different directories aren't cleaned anymore. What's the plan here? Do nothing and wait until we abandon plain makefile support? A pending fix? I would like to have at least a comment explaining what's up.

**Comment by ronald-cron-arm** [comment]
I try to not break anything. I am going to fix that in the next PR.

File: library/Makefile
Comment (ronald-cron-arm) [reply library/Makefile:402]: I try to not break anything. I am going to fix that in the next PR.

Comment (gilles-peskine-arm) [reply library/Makefile:402]: Ok, thanks. I noticed because of a conflict with https://github.com/Mbed-TLS/mbedtls/pull/9293, so your fix would conflict with that too.

File: tf-psa-crypto/core/common.h
Comment (gilles-peskine-arm) [inline tf-psa-crypto/core/common.h:1]: > The common.h file is copied instead of
being just moved as eventually they
will be different in mbedtls and
TF-PSA-Crypto. Some parts of it can be
shared though but this will be done later,
probably when adding the CMake build
system in tf-psa-crypto.

_sigh_ Why? Why isn't `library/common.h` just `#include "core/common.h"`? Or just omit `library/common.h` since we don't actually have anything to put there? Why are we spending effort on keeping two files identical? And how will that work when they're no longer in the same repository?

**Pull request closed without merging**
